integrations/tensorflow/e2e/keras/lstm_static_test.py

```diff
@@ -42,7 +43,6 @@ def __init__(self):
             self.m.call)
 
 
-@tf_test_utils.compile_module(LstmStaticModule, exported_names=["predict"])
 class LstmStaticTest(tf_test_utils.TracedModuleTestCase):
```

Comment:
Can we make this look like:
```
compiled_modules = tf_test_utils.compile_tf_module(LstmStaticModule, exported_names=["predict"])

class LstmStaticTest(tf_test_utils.TracedModuleTestCase):
  def test_lstm(self):
    def predict(module):
      ...
    self.compare_backends(compiled_modules, predict)
```

Then we don't have to depend on spooky action-at-a-distance with globals.

Reply:
I'm not sure how to do that off the top of my head. The problem is that absl flags aren't parsed when that line runs, so we won't be able to get the target backends that we need to compile.

Reply:
ok, then can we do this then?

```
compiled_modules = None

class LstmStaticTest(tf_test_utils.TracedModuleTestCase):
  def test_lstm(self):
    def predict(module):
      ...
    self.compare_backends(compiled_modules, predict)

def main():
  global compiled_modules
  compiled_modules = tf_test_utils.compile_tf_module(LstmStaticModule, exported_names=["predict"])
```

Reply:
actually, can we just do the compilation in the setUp method of the test case module?

Reply:
Do you mean `setUp` or `setUpClass`. We could conditionally compile the modules in `setUp` based on whether or not we've already compiled them, but at that point we might as well just keep using `setUpClass`.

And yes we could. That would mean that we have two decorators that can be applied to the test case (`@compile_tf_module` and `@compie_signature_def_saved_model`). The test case would need to know how to conditionally process and use the attributes each decorator assigns to its `cls` (`_module_class`, and `_exported_names` in one case, and `_saved_model_dir`, `_saved_model_tags`, `_module_name`, `_exported_name/_signature` in the other) to compile the data that it's passed. We would still need to use global variables to store the `CompiledModule`s because of #2900 however.

My motivation for removing the test case decorator is that the test case would no longer need to know the specific process for taking two sets of inputs and producing a compiled modules. The code for making the test case do that wouldn't be too horrendous though, assuming we won't need to add a third compilation path at some point.

What do you think? I doubt I've exhausted all possibilities for how we could do this here.

Reply:
I guess I meant setUpClass (wasn't familiar with that distinction). I don't think we even need decorators (those just make it a bit more obfuscated what is happening). Can we do it low-tech like this?

```
class LstmStaticTest(tf_test_utils.TracedModuleTestCase):
  def setUpClass(self):
    self.compiled_modules = tf_test_utils.compile_tf_module(LstmStaticModule, exported_names=["predict"])
  def test_lstm(self):
    def predict(module):
      ...
    self.compare_backends(self.compiled_modules, predict)
```

Reply:
I like that. It adds a fair bit of boilerplate, but I think the clarity is worth it:

```python
class SimpleArithmeticTest(tf_test_utils.TracedModuleTestCase):

  @classmethod
  def setUpClass(cls) -> None:
    # We still need to use global variables because of #2900.
    tf_test_utils.compile_tf_module(SimpleArithmeticModule)
    # This needs to come _after_ the above command so that we can keep the 
    # AttributeError checking
    super().setUpClass()

  def test_simple_mul(self):

    def simple_mul(module):
      a = np.array([1., 2., 3., 4.], dtype=np.float32)
      b = np.array([400., 5., 6., 7.], dtype=np.float32)
      c = module.simple_mul(a, b)
      module.simple_mul(a, c)

    self.compare_backends(simple_mul)
```

Unfortunately we still can't put the modules on the TestCase class though.

Reply:
> Unfortunately we still can't put the modules on the TestCase class though.

Why not?

Reply:
Python garbage collection does not work correctly when the modules are on the class (literally on `cls`) which causes test failures on nvidia GPUs (see #2659).

We initially fixed this in `setUp` by recompiling the modules for every trace, but that made the tests take a lot longer. We tried to fix this by only compiling in the first `setUp` call, but everything we tried to get the modules to stay on `self` was thwarted by something inherited from the `TestCase` hierachy.

That said I did some digging and figured out how to override `__init__` without breaking things, and I _think_ that it will probably still work on GPU (I'll need to trigger some presubmits internally though to test it). That would look like this:

```python
class SimpleArithmeticTest(tf_test_utils.TracedModuleTestCase):

  def __init__(self, methodName="runTest"):
    super(SimpleArithmeticTest, self).__init__(methodName)
    self._modules = tf_test_utils.compile_tf_module(SimpleArithmeticModule)

  def test_simple_mul(self):

    def simple_mul(module):
      a = np.array([1., 2., 3., 4.], dtype=np.float32)
      b = np.array([400., 5., 6., 7.], dtype=np.float32)
      c = module.simple_mul(a, b)
      module.simple_mul(a, c)

    self.compare_backends(simple_mul, *self._modules)
```

With this modification to `compare_backends`
```python
  def compare_backends(self, trace_function: Callable[[TracedModule], None],
                       ref_module: tf_utils.CompiledModule,
                       tar_modules: Sequence[tf_utils.CompiledModule],
                       artifacts_dir: str) -> None:
```

The name `_modules` is kind of a lie, but it's a less verbose than writing everything out, and serves the purpose of documenting that `compare_backends` is comparing `simple_mul` on the modules we compile.

Reply:
Sweet. That looks great.

nit: You might want to make self._modules be a namedtuple and have compared_backends take that tuple (not sure if the deps work out on having that tuple visible everywhere). Just to avoid the *self._modules which is a bit of a footgun (people can forget the *, etc.)

Reply:
Sounds good. I was considering a dict but they're kinda ugly.

How would you feel about using dataclasses? They were introduced in 3.7, but are backported to 3.6 via `python3 -m pip install dataclasses`. (TensorFlow now lets us use 3.8 so I would prefer to assume that everyone has a copy of that installed, but that'd be a bigger discussion).

Reply:
Oh cool. Didn't know about dataclasses. They look cool but I'd rather just keep it as simple as possible for now (anything that involves thinking about dependencies or available versions I don't consider "simple" :) ).

Reply:
Done! :)